File: Sort_OC/Sort_OC/QuickSort.m
//
//  QuickSort.m
//  Sort_OC
//
//  Created by lu hongliang on 2019/4/10.
//  Copyright © 2019 lu hongliang. All rights reserved.
//

#import "QuickSort.h"

@implementation QuickSort

- (void)sort:(NSMutableArray *)array :(NSNumber *)start :(NSNumber *)end
{
    if (start.integerValue > end.integerValue) {
        return;
    }
    if (array.count <= 0) {
        return;
    }
    
    NSInteger index = start.intValue;
    for(int i = start.intValue; i < end.intValue; i++) {
        if (((NSNumber *)array[i]).integerValue < ((NSNumber *)array[end.integerValue]).integerValue) {
            if (index != i) {
                [self swap:array :@(index) :@(i)];
            }
            index ++;
        }
    }
    
    if (index != end.integerValue) {
        [self swap:array :@(index) :end];
    }
    
    [self sort:array :start :@(index - 1)];
    [self sort:array :@(index + 1) :end];
}

- (void)swap:(NSMutableArray *)array :(NSNumber *)one :(NSNumber *)two
{
    NSNumber *tmp = array[one.intValue];
    array[one.intValue] = array[two.intValue];
    array[two.intValue] = tmp;
}

@end

File: Sort/Sort/QuickSort.swift
//
//  QuickSort.swift
//  Sort
//
//  Created by honglianglu on 12/05/2018.
//  Copyright © 2018 cc. All rights reserved.
//

import UIKit

class QuickSort: Sort {
    func sort(_ nums: inout [Int],_ first: Int, _ end: Int) {
        if first > end {
            return
        }
        var index = first
        for i in first...end {
            if nums[i] < nums[end] {
                if i != index {
                    nums.swapAt(i, index)
                }
                index += 1
            }
        }
        if index != end {
            nums.swapAt(index, end)
        }
        self.sort(&nums, first, index - 1)
        self.sort(&nums, index + 1, end)
    }
}
